### ISSUE TITLE
How to write custom sampler/dataloader for large dataset loading from disk

### ISSUE THREAD

**Issue by ray38** [skorch-dev/skorch#753]
Hi all! As I'm relatively new to skorch, I'm having trouble optimizing this workflow to train data huge datasets prepared as a series of lmdb files (each contains several thousands of data points). I think I'm 80% there now, but got stuck writing a custom sampler for the training dataset, and would really appreciate your help.  sorry in advance that this post might be a little long.

As mentioned, data is prepared as a series of lmdb file, so we can access and load specific datapoints from disk during training, tho it's super slow as it's random access, So I tried to optimize it by caching lmdb files in memory sequentially during training, and that decrease the time needed for each epoch by more than 50x.

Here is the relevant part of the dataset class I used to achieve this:

```
class LMDBDatasetPartialCache(Dataset):

    def __init__(
        self,
        db_paths,
    ):
        self.db_paths = db_paths
        self.envs = []
        self.keys_list = []
        self.length_list = []

        for db_path in self.db_paths:
            temp_env = self.connect_db(db_path)
            self.envs.append(temp_env)
            self.keys_list.append(
                [f"{j}".encode("ascii") for j in range(temp_env.stat()["entries"])]
            )
            with temp_env.begin(write=False) as txn:
                temp_length = pickle.loads(txn.get("length".encode("ascii")))
                self.length_list.append(temp_length)

        self._keylen_cumulative = np.cumsum(self.length_list).tolist()
        self.total_length = np.sum(self.length_list)


    def __len__(self):
        return self.total_length

    def __load_dataset__(self, db_idx):
        dataset = []
        with self.envs[db_idx].begin(write=False) as txn:
            for idx in range(self.length_list[db_idx]):
                data = txn.get(self.keys_list[db_idx][idx])
                data_object = pickle.loads(data)
                dataset.append(data_object)

        self.loaded_db_idx = db_idx
        self.loaded_dataset = dataset

    def __getitem__(self, idx):
        db_idx = bisect.bisect(self._keylen_cumulative, idx)
        if db_idx != 0:
            el_idx = idx - self._keylen_cumulative[db_idx - 1]
        else:
            el_idx = idx

        if db_idx != self.loaded_db_idx:
            self.__load_dataset__(db_idx)

        return self.loaded_dataset[el_idx]

```

(let's assume there are M lmdb files with N datapoints in total)

this was used in combination with in-order splitter (first x percent as training, last 1-x percent as validation, instead of randomized) developed here: https://github.com/skorch-dev/skorch/issues/747#issuecomment-802242934
and setting `iterator_train__shuffle=False`. Hence, during training, it will go from data point 1 to N sequentially, which means it will cache the lmdb files from 1 to M in memory sequentially for faster access.

This is all good until I realized randomizing the feed sequence of training data is crucial to getting good model... this would mean setting `iterator_train__shuffle=False`, but it will break the whole cache scheme since we will be constantly switching between different lmdb files instead of processing them sequentially. Therefore I think a custom sampler/dataloader is the way to go but not really sure how to achieve it. 

The best way of having a pseudo-shuffle but keep cache scheme valid is to shuffle the sequence of lmdb files, and shuffle the sequences within each lmdb file for each epoch, so probably something like this?

```
class DatapointSampler(Sampler):
    
    def __init__(self, length_list):
        self.num_datasets = len(dataset)
        self.length_list = length_list
    def __iter__(self):
        datapoint_order = []

        dataset_order = [i for i in torch.randperm(self.num_datasets)]
        current_total_counter = 0
        for dataset_idx in dataset_order:
            datapoint_order += [i+current_total_counter for i in torch.randperm(self.length_list[dataset_idx])]
            current_total_counter += self.length_list[dataset_idx]

        return datapoint_order
```
(Please let me know if there's logic/syntax error, or there's better ways of doing it)

What I'm not clear is how to use this with the in-order splitter, as this sampler as-is will be drawing datapoints from the whole dataset (training + valid), which is totally undesired right? do I actually need custom dataloaders for train and valid to make this work? 

Thanks a lot for your help, any suggestion will be greatly appreciated! and please let me know if you need me to clarify anything.

**Comment by BenjaminBossan** [comment]
So I cannot say I understood everything with the given code and explanation. Thus I cannot tell you if your caching and sampling logic are sound. I assume that you already exhausted other options like using LRU cache or multiple jobs and found that insufficient/inadequate to solve your problem.

> What I'm not clear is how to use this with the in-order splitter, as this sampler as-is will be drawing datapoints from the whole dataset (training + valid), which is totally undesired right? do I actually need custom dataloaders for train and valid to make this work?

The logic works like this:

1. You pass your data to skorch
2. data is packaged into a `Dataset` if it isn't already
3. the `Dataset` is split into two `Subset`s if a `train_split` is being used, "train" and "valid"
4. each of the two sets gets their own `DataLoader` inside the `get_iterator` method
5. each `DataLoader` has its own sampler (`SequentialSampler` is the PyTorch default)

If I understand your question correctly, your custom sampler therefore doesn't interfere with the train/valid split, since it only ever gets to see the training data. If my understanding is incorrect, I would ask you to provide a bit more information.

**Comment by ray38** [comment]
@BenjaminBossan Thanks a lot for your reply!
I guess the question is how should I initialize the sampler
i.e. if I set `iterator_train__sampler=mysampler`, where the `length_list` is the list lengths of all lmdb files provided to the dataset. This means the indices that the sampler sees are not split according to the splitter. so I'm afraid it will draw datapoints (by their indices) from the whole pool (train + val) instead of val only. 

maybe I should just initialize the sampler with the splitter (or split percentage) then? it's means to always be used with the in-order splitter anyway, and with the splitter I can easily determine the max index of the train set, and only draw randomly up to that during training?

**Comment by BenjaminBossan** [comment]
If I don't misunderstand, it should be impossible for the sampler to "accidentally" pick an index that it's not supposed to see. E.g. let's assume we have a dataset of 1000 samples. Using `InOrderSplit`, `[0:800]` are assigned as train and `[800:1000]` as valid. These indices are used to create the `Subset`s. Now if you would try to access index 800, 801 etc. from the train `Subset`, you would get an index error. For the valid `Subset`, index 0 will map to 800, 1 to 801 etc., starting from 200, there is again an index error. Therefore, if your splitter tried to access data it's not supposed to see, you would get an error.

Now again, I didn't get everything about your use case, so maybe there is something happening that violates my assumption. From what I perceive, what you would need to do in that case is to initialize the sampler _after_ the datasets have been split, which means you cannot pass the splitter as `iterator_train__sampler=mysampler`. What I would suggest is to override the `get_iterator` method:

https://github.com/skorch-dev/skorch/blob/afa697ed1b16b0c0e6b661ed0018d4a332da90fb/skorch/net.py#L1353-L1390

When you look at it, it should be quite straightforward what's happening. For your case, the interesting part is that the method receives the initialized dataset and returns the `DataLoader`. Therefore, you could initialize your sampler right there, with access to the dataset.

> maybe I should just initialize the sampler with the splitter (or split percentage) then? it's means to always be used with the in-order splitter anyway, and with the splitter I can easily determine the max index of the train set, and only draw randomly up to that during training?

That would work but it sounds brittle. E.g. if you increase or decrease the size of your dataset, you would need to remember to change the splitter as well. It's not impossible, just a potential source of hard to spot bugs.

**Comment by BenjaminBossan** [comment]
@ray38 Any progress?

**Comment by ray38** [comment]
@BenjaminBossan Yes! Thanks a lot for your help
The solution I took was straightforward, just initialize the sampler with the fraction of validation datapoints. with the inorder splitter, it worked out perfectly in the end. Here's the code for the sampler just in case anybody is interested:

```
class PartialCacheSampler(Sampler):
    def __init__(self, length_list, val_frac):
        len_cumulative = np.cumsum(length_list)
        len_dataset = np.sum(length_list)
        len_val = int(len_dataset * val_frac)
        len_train = len_dataset - len_val
        for i, cum_len in enumerate(len_cumulative):
            if cum_len >= len_train:
                self.length_list = length_list[: i + 1]
                self.length_list[-1] -= cum_len - len_train
                break

        self.num_datasets = len(self.length_list)
        self.start_idx_list = [0] + np.cumsum(self.length_list).tolist()
        self.total_length = np.sum(self.length_list)

    def __iter__(self):
        datapoint_order = []
        dataset_order = torch.randperm(self.num_datasets).tolist()
        for dataset_idx in dataset_order:
            start_idx = self.start_idx_list[dataset_idx]
            datapoint_order += [
                i + start_idx
                for i in torch.randperm(self.length_list[dataset_idx]).tolist()
            ]
        return iter(datapoint_order)
```

Thanks a lot again, and let me close this issue
